From 11329d5d00bf6d5c6eb666593181e50a3056b46c Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Thu, 10 Nov 2022 14:44:24 +0200
Subject: [PATCH] [SW-110215] pcie_pif: implement reset control

The new control register implements a reset enable bit which enables or
disables the FLR reset funcitonality.
To avoid any doubt, only the reset functionality itself is affected, the
rest of the FLR flow remains unaffected by this bit.
Implement reset functionality control during the FLR handling.

Change-Id: Iaa465605d168f22bbaaa6840401f126aa424ab27
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/app/pcie_pif/pcie_pif.c b/app/pcie_pif/pcie_pif.c
index da57c4b..b8851eb 100644
--- a/app/pcie_pif/pcie_pif.c
+++ b/app/pcie_pif/pcie_pif.c
@@ -81,6 +81,11 @@
 	flr_report_status(STAT_FLR_DONE_REPORTED);
 }
 
+static inline bool pif_flr_rst_en(void)
+{
+	return !!(regmap_readl(pif_arc.control_reg) & PIF_CTRL_FLR_RST_EN);
+}
+
 static void wait_for_secondary_die(void)
 {
 	while (!regmap_readl(pif_arc.psoc2pif_die_ready)) {
@@ -159,6 +164,9 @@
 
 static void trigger_flr_reset(void)
 {
+	if (!pif_flr_rst_en())
+		return;
+
 	/* make sure HW ACK flag is cleared */
 	regmap_writel(0, pif_arc.psoc2pif_flr_done);
 
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index cc52c61..0f86d9d 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -415,6 +415,7 @@
 		pif_arc.timestamp_cnt = mmD0_PIF_ARC_FLR_TIMESTAMP_CNT;
 		pif_arc.error_reg = mmD0_PIF_ARC_ERROR;
 		pif_arc.status_reg = mmD0_PIF_ARC_STATUS;
+		pif_arc.control_reg = mmD0_PIF_ARC_CONTROL;
 	} else {	/* DIE1 */
 		pif_arc.pf_flr_active = mmD1_PIF_ARC_PF_FLR_ACTIVE;
 		pif_arc.pf_flr_done = mmD1_PIF_ARC_PF_FLR_DONE;
diff --git a/include/device.h b/include/device.h
index e6fe6c5..dc036f7 100644
--- a/include/device.h
+++ b/include/device.h
@@ -98,6 +98,7 @@
 	u32 timestamp_cnt;
 	u32 error_reg;
 	u32 status_reg;
+	u32 control_reg;
 };
 
 struct device_pll {
